samza-core/src/main/java/org/apache/samza/zk/BarrierForVersionUpgrade.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.zk;
+
+import java.util.List;
+
+
+/**
+ * Interface for a barrier - to allow synchronization between different processors to switch to a newly published
+ * JobModel.
+ */
+public interface BarrierForVersionUpgrade {
+  /**
+   * Barrier is usually started by the leader.
+   * @param version - for which the barrier is started.
+   * @param processorsNames - list of processors available at the time of the JobModel generation.
+   */
+  void startBarrier(String version,  List<String> processorsNames);
```

Comment:
Ideally speaking, what we have implemented is some kind of computation latch that only synchronizes the beginning of computation. I would actually expect the functions in barrier to be "create" (done by the leader, which simply creates the znode), "enter" (every processor joins the barrier znode) and "leave" (every processor leaves the barrier znode after computation is complete. See [example](https://zookeeper.apache.org/doc/r3.3.6/zookeeperTutorial.html#sc_barriers)

Having said that, I agree with Fred's suggestion about removing the word "barrier" from the function name. It seems redundant as the interface already reads BarrierForVersionUpgrade. It will also be nice to use "start" instead of "create".